backend/server.js

```diff
@@ -202,17 +207,138 @@ sbp('sbp/selectors/register', {
+      // (attachments) and other contracts. Removing a single contract could
+      // therefore result in a large number of contracts being deleted. For
+      // example, in Group Income, deleting an identity contract will delete:
+      //   - All groups created by that contract
+      //       - This includes files like the group avatar
+      //       - And also all chatrooms
+      //           - And all attachments in chatrooms
+      //   - All DMs created by that contract
+      //       - And all attachments
+      const resourcesKey = `_private_resources_${cid}`
+      const resources = await sbp('chelonia.db/get', resourcesKey)
+      if (resources) {
+        await Promise.allSettled(resources.split('\x00').map((resourceCid) => {
+          const parsed = parseCID(resourceCid)
+
+          if (parsed.code === multicodes.SHELTER_CONTRACT_DATA) {
+            return sbp('chelonia.persistentActions/enqueue', ['backend/deleteContract', resourceCid])
+          } else if (parsed.code === multicodes.SHELTER_FILE_MANIFEST) {
+            return sbp('chelonia.persistentActions/enqueue', ['backend/deleteFile', resourceCid])
+          } else {
+            console.warn({ cid, resourceCid, code: parsed.code }, 'Resource should be deleted but it is of an unknown type')
+          }
+
+          return undefined
+        }))
+      }
+      await sbp('chelonia.db/delete', resourcesKey)
+
+      // Next, loop through all the events in the contract and delete them,
+      // starting with the most recent ones.
+      // If the deletion process is interrupted, parts of the contract will
+      // still be able to be synced, but won't be to write to it (due to
+      // latestHEADinfo not being deleted).
+      const latestHEADinfo = await sbp('chelonia/db/latestHEADinfo', cid)
+      if (latestHEADinfo) {
+        for (let i = latestHEADinfo.height; i > 0; i--) {
+          const eventKey = `_private_hidx=${cid}#${i}`
+          const event = await sbp('chelonia.db/get', eventKey)
+          if (event) {
+            await sbp('chelonia.db/delete', event)
+            await sbp('chelonia.db/delete', eventKey)
+          }
+        }
+        await sbp('chelonia/db/deleteLatestHEADinfo', cid)
+      }
+
+      // Then, delete all KV-store values associated with this contract
+      const kvIndexKey = `_private_kvIdx_${cid}`
+      const kvKeys = await sbp('chelonia.db/get', kvIndexKey)
+      if (kvKeys) {
+        await Promise.all(kvKeys.split('\x00').map((key) => {
+          return sbp('chelonia.db/delete', `_private_kv_${cid}_${key}`)
+        }))
+      }
+      await sbp('chelonia.db/delete', kvIndexKey)
+
+      await sbp('chelonia.db/get', `_private_cid2name_${cid}`).then((name) => {
+        if (!name) return
+        return Promise.all([
+          sbp('chelonia.db/delete', `_private_cid2name_${cid}`),
+          appendToOrphanedNamesIndex(name)
+        ])
+      })
```

Comment:
Question: what about the `name=` mapping? Should anything happen to that? (e.g. be set to `''`?)

Reply:
It could, but I don't see the need to do so.